simulated type fixes

kuceb opened the pull request:
<!-- Thanks for contributing!
Read our contribution guidelines here: 
https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md -->

- [x] type validation during every character input, type following activeElement/focus
fix #2240
- [x] fix not borrowing property getter for `maxLength` during type validation
fix #4587
- [x] allow targeting any focusable element during `cy.type`
fix #2166
fix #3661
- [x] not updating value in `datetime-local` input
fix #2613
- [x] allow typing minus(`-`) sign into number field if entire field is selected
fix #4767 
- [x] fire input event when {enter} inserts newline
fix #3405
- [x] don't input text when non-shift modifier is pressed
fix #5424
- [x] fire 'code' property on keyboard events
fix #3722


- [x] ~use KeyboardEvent constructor instead of Event from app frame~ _forcing certain property values on the event during and after construction is not stable (does not work in chrome 63)_
- [x] use host contenteditable as focusable element when targeting element inside contenteditable (should help with rich text editors)
- [x] remove need to intercept `select` command and keep cypress cursor state on form elements



### Tasks
- [x] waiting on #3030 to be merged
- [x] wait on #5383 moving many typescript specific changes from this PR into there to make this diff manageable
- [x] wait on #5382  eslint config changes in separate PR (typescript-eslint changes)
- [x] fix error message for `cy.type`, use definition of `isFocusable`

### UX Changes:
error message on not typeable element
- before:
- after:
![19-10-18_17:52::12](https://user-images.githubusercontent.com/14625260/67130753-57339280-f1d0-11e9-9bd5-0a687711a063.png)

error message on not clearable element
- before:
- after:
![image](https://user-images.githubusercontent.com/14625260/67130739-4daa2a80-f1d0-11e9-81f1-639a2d52e52a.png)

type follows focus




### Pre-merge Tasks

<!-- The following tasks must be completed before a PR can be merged.
You can delete tasks if they are not applicable to the PR changes. -->

- [x] Have tests been added/updated for the changes in this PR?
- [x] Has the original issue been tagged with a release in ZenHub? <!-- (internal team only)-->

kuceb committed:
rename mouse/keyboard

kuceb committed:
apply changes on this branch with rename

kuceb committed:
temp 07/18/19 [skip ci] type_spec minor changes

kuceb committed:
cleanup type_spec, allow unused-vars for args

kuceb committed:
fix missing return

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
temp 07/19/19 [skip ci]

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
fix(driver:keyboard): fix typing negative numbers with existing text selection

kuceb committed:
temp 07/22/19 [skip ci]

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
temp 07/22/19 [skip ci]

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
remove accidental spec, use beforeEach visit in click_spec

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
Merge remote-tracking branch 'origin/issue-2957' into issue-2957

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
add tests for cy.trigger changes

kuceb committed:
re-run build

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
extract simulated type from native events

kuceb committed:
most type_spec passing until tab

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
add typedefs

kuceb committed:
temp 08/08/19 [skip ci]

kuceb committed:
fix type_spec, click_spec

kuceb committed:
fix console table event table logging

kuceb committed:
Merge branch 'develop' into issue-2957

kuceb committed:
fix spec utils

kuceb committed:
temp 08/12/19 [skip ci]

kuceb committed:
Merge remote-tracking branch 'origin/develop' into simulated-type-fixes

kuceb committed:
large changes to selection.ts, working type_spec, eslint changes

kuceb committed:
fix elements util, cleanup debugging code

kuceb committed:
fix linting rules

kuceb committed:
fix type_spec in electron by using plain Event contructor

kuceb committed:
fix invalid clicking-into-iframe spec

kuceb committed:
address review, cleanup

kuceb committed:
use webpack-preprocessor (required for typescript files)

kuceb committed:
fix invalid connectors_spec

kuceb committed:
chore(lint): fix linting, since changed rules

kuceb committed:
fix not borrowing property getter for maxlength

kuceb committed:
Merge remote-tracking branch 'origin/develop' into simulated-type-fixes

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

jennifer-shehane commented on packages/driver/src/cy/keyboard.ts:
```diff
@@ -0,0 +1,1156 @@
+    // const isFocusable = $elements.isFocusable($dom.wrap(el))
+    // const isTextLike = $elements.isTextLike(el)
+
+    // const isTypeableButNotTextLike = !isTextLike && isFocusable
```
remove this commented out code

jennifer-shehane commented on packages/driver/src/cy/keyboard.ts:
```diff
@@ -0,0 +1,1156 @@
+    return {}
+  }
+
+  if (!isFocusable && !isTextLike) {
```
These 2 if statements are confusing. We're going to throw an error if `!isFocusable`, then we are further distinguishing the error shown by whether `isTextLike` or not. So, maybe:

```suggestion
  if (!isFocusable) {
   const node = $dom.stringify($el)
   const errMsg = isTextLike ? 'not_actionable_textlike' : 'not_on_typeable_element' 

    $utils.throwErrByPath(`type.${errMsg}`, {
      onFail,
      args: { node },
    })
 
  }
```

jennifer-shehane commented on packages/driver/src/cy/commands/actions/type.js:
```diff
@@ -65,520 +65,538 @@ module.exports = function (Commands, Cypress, cy, state, config) {
-      const isMonth = $dom.isType(options.$el, 'month')
-      const isWeek = $dom.isType(options.$el, 'week')
-      const hasTabIndex = $dom.isSelector(options.$el, '[tabindex]')
+    // const verifyElementForType = ($el) => {
```
There's a lot of commented out code in this file too, can this be removed??

jennifer-shehane commented on packages/driver/src/cypress/error_messages.coffee:
```diff
@@ -915,6 +924,7 @@ module.exports = {
     invalid_time: "Typing into a time input with #{cmd('type')} requires a valid time with the format 'HH:mm', 'HH:mm:ss' or 'HH:mm:ss.SSS', where HH is 00-23, mm is 00-59, ss is 00-59, and SSS is 000-999. You passed: {{chars}}"
+    invalid_dateTime: "Typing into a datetime input with #{cmd('type')} requires a valid datetime with the format 'yyyy-MM-ddThh:mm', for example '2017-06-01T08:30'. You passed: {{chars}}"
     multiple_elements: "#{cmd('type')} can only be called on a single element. Your subject contained {{num}} elements."
     not_on_typeable_element: """
```
refactor these error messages like we said - listing out what we consider 'isFocusable' elements - throw a more specific error message if the element is `[disabled]` and then list the other elements we check (with disabled excluded) for the error message if not already caught.

eps1lon replied:
This may seem very naive since I've spent barely any time with the implementation. I'm concerned about the kind of tests the current API allows with regard to accessibility:

Why do you even check focusability? The user agent already takes care of this and any custom logic you have will require maintenance. Simply forwarding `type` to something like `document.activeElement.dispatchEvent` seems like a simpler implementation that isn't subject to any confusion as to which elements are able to receive keyboard events.

The testing API shouldn't allow specifying the element that receives the keyboard event in the first place (or only when enabling a nasty sounding flag like `unsafe` or `force`). Otherwise you can construct test cases that are unreachable by a  real user.

jennifer-shehane replied:
>The user agent already takes care of this

I'm not sure what you mean by this. As browser automation software, Cypress has to programmatically automate everything, there is not a real user clicking, etc. Knowing what allows a focus event is a necessary part of any browser automation software. You're maybe thinking that Native browser events support would be better? https://github.com/cypress-io/cypress/issues/311 We agree and this is one of the requirements for moving forward with that.

eps1lon replied:
Hm maybe I'm misunderstanding something about the architecture but you do use the DOM provided by Chrome or electron no?

These user agents already prevent focus on an element that can't receive focus e.g.

```html
<main>try focusing me</main>
```

```js
console.assert(document.activeElement === document.body);

const main = document.querySelector('main');
main.focus();

console.assert(document.activeElement === document.body);
```
does already work. You do not have to implement any custom logic to only allow focus on e.g. elements with tabindex or `<button />` etc. So if you do have access to the DOM of a user agent such as chrome or electron there's no need for a custom focus filter.

Bkucera replied:
> Why do you even check focusability? The user agent already takes care of this and any custom logic you have will require maintenance.

@eps1lon this is a very common misconception that the user agent will perform default actions (inserting text, moving focus, moving cursor) by using `dispatchEvent`. This isn't true, and we must implement the default actions ourselves. The only case where the browser will react to a synthetic event sent through `dispatchEvent` is a checkbox can be toggled with the `click` event.

Bkucera replied:
@eps1lon also, when the browser is out of focus, `document.activeElement` will be updated, but focus events are queued until the window focus is regained. Because of this we need to "polyfill" focus events in some cases

eps1lon replied:
> this is a very common misconception that the user agent will perform default actions (inserting text, moving cursor) by using dispatchEvent.

I understand that. I'm specifically talking about `HTMLElement.focus()` not `dispatchEvent(new FocusEvent())`. See https://github.com/cypress-io/cypress/pull/4870#discussion_r328651245

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
temp 09/25/19 [skip ci]

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
add test for clicking checkbox, cleanup click_spec

kuceb committed:
document mouse click and mouse move algo, adjust mouse move, add test for recursive movement

brian-mann committed:
Merge branch 'develop' into issue-2957

brian-mann committed:
cleanup / formatting

brian-mann committed:
scope the getCommandLogWithText to the active runnable to prevent leakage

brian-mann committed:
cleanup tests, make values static so tests pass consistently

- use attach listener alias APIs
- remove magic number constants
- fix some incoherent html / script tag stuff

brian-mann committed:
begin implementation for returning fromAutWindow

- restore original fromWindow algorithm
- begin writing fromAutWindow for calculating red hitbox coordinates
which are layered on the AUT

brian-mann committed:
move describe block with the others

brian-mann committed:
refactor actions to DRY up duplicated logic between click/dblclick/rightclick

- tighten up the mouse.js method names, name them consistently

brian-mann committed:
remove comment

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
Merge remote-tracking branch 'origin/issue-2957' into issue-2957

kuceb committed:
fix iframe coords and test

kuceb committed:
Merge branch 'issue-2957' into simulated-type-fixes

kuceb committed:
temp 10/06/19 [skip ci]

jennifer-shehane committed:
Merge branch 'develop' into issue-2957

kuceb committed:
rename fromDocCoords->fromWindowCoords, fix type_spec ux test

kuceb committed:
Merge remote-tracking branch 'origin/issue-2957' into issue-2957

kuceb committed:
temp 10/07/19 [skip ci]

kuceb committed:
Merge branch 'issue-2957' into simulated-type-fixes

jennifer-shehane commented:
Need to fix some conflicts. 😢

kuceb committed:
fix after merge from mouse-fixes branch / develop

kuceb committed:
Merge remote-tracking branch 'origin/develop' into simulated-type-fixes

kuceb committed:
temp 10/08/19 [skip ci]

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
Merge branch 'issue-2957' into simulated-type-fixes

kuceb committed:
Revert "temp 10/08/19 [skip ci]"

This reverts commit 92b73f6107808a35857cbaa8d08c28f00eb2c33e.

kuceb committed:
remove only, fix contenteditable test

kuceb committed:
add test for datetime-local inputs

kuceb committed:
add webpack devdep

kuceb committed:
fix debugging code

kuceb committed:
fix force:true on hidden input

kuceb committed:
up timeout for test in ci

kuceb committed:
Merge remote-tracking branch 'origin/develop' into simulated-type-fixes

brian-mann committed:
tighten up method, remove unnecessary arg

brian-mann committed:
Merge remote-tracking branch 'origin/issue-2957' into issue-2957

brian-mann committed:
add debug logic for retries to console.error() non cypress errors

brian-mann committed:
renamed fromWindow -> fromElWindow, and fromViewport -> fromElViewport

brian-mann committed:
handle calculating the fromElWindow coordinates inline

brian-mann committed:
extract out function for clarity, receive consistent arg names

brian-mann committed:
cleanup, add lots of comments

brian-mann committed:
Merge branch 'develop' into issue-2957

# Conflicts:
#	packages/driver/src/cy/commands/screenshot.coffee
#	packages/driver/test/cypress/integration/dom/coordinates_spec.coffee
#	packages/driver/test/cypress/integration/e2e/dom_hitbox.spec.js

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
Merge branch 'issue-2957' into simulated-type-fixes

kuceb committed:
fix display specific test

kuceb committed:
fix type follow focus when redirect on focus

kuceb committed:
cleanup new follow focus test

kuceb committed:
upgrade eslint-plugin-dev to 5.0.0

kuceb committed:
use typescript in driver

kuceb committed:
more [lint fixes]

kuceb committed:
Merge branch 'develop' into upgrade-eslint-plugin-dev

kuceb committed:
Merge branch 'upgrade-eslint-plugin-dev' into ts-in-driver

kuceb committed:
Merge remote-tracking branch 'origin/develop' into ts-in-driver

kuceb committed:
cleanup minor

kuceb committed:
remove unneeded file

kuceb committed:
mild refactor webpack configs

kuceb committed:
Merge branch 'develop' into ts-in-driver

kuceb committed:
fix elements.ts

kuceb committed:
Merge remote-tracking branch 'origin/ts-in-driver' into ts-in-driver

kuceb committed:
fix isTextLike signature change

kuceb committed:
fix isType check, fix type_spec

kuceb committed:
actually use webpack-preprocessor

kuceb committed:
Merge branch 'ts-in-driver' into simulated-type-fixes

kuceb committed:
fix isType domUtil

kuceb committed:
fix connectors_spec

kuceb committed:
cleanup connectors_spec

kuceb committed:
address feedback- cleanup, add comments, refac

kuceb committed:
answer: no, need index.js

kuceb committed:
Merge branch 'ts-in-driver' into simulated-type-fixes

kuceb committed:
Merge remote-tracking branch 'origin/develop' into ts-in-driver

kuceb committed:
Merge branch 'ts-in-driver' into simulated-type-fixes

kuceb committed:
cleanup after driver typescript PR

kuceb committed:
fix keyboard getActiveModifiers

Bkucera commented on packages/driver/src/cy/retries.coffee:
```diff
@@ -7,6 +7,10 @@ $utils = require("../cypress/utils")
 create = (Cypress, state, timeout, clearTimeout, whenStable, finishAssertions) ->
   return {
     retry: (fn, options, log) ->
+      ## FIXME: remove this debugging code
```
remove this before merge

kuceb committed:
remove debugging code, cleanup

brian-mann commented on packages/driver/src/cy/keyboard.ts:
```diff
@@ -1,44 +1,103 @@
+    [key in KeyEventType]?: boolean;
+  }
+}
+// const $Cypress = require('../cypress')
```
comment

Bkucera replied:
fixed

brian-mann commented on packages/driver/src/cy/keyboard.ts:
```diff
@@ -47,680 +106,1010 @@ const toModifiersEventOptions = (modifiers) => {
+const fromModifierEventOptions = (eventOptions: {
+  [key: string]: string
+}): keyboardModifiers => {
+  const modifiers = _.pickBy(
```
tighten up

brian-mann replied:
`_.pickBy` uses `_.identity` as the default

brian-mann commented on packages/driver/src/cy/keyboard.ts:
```diff
@@ -47,680 +106,1010 @@ const toModifiersEventOptions = (modifiers) => {
+    }
+
+    return Promise.each(typeKeyFns, (fn) => {
+      return Promise.try(() => {
```
tighten up

brian-mann commented on packages/driver/src/dom/elements.ts:
```diff
@@ -23,7 +23,19 @@ const focusableSelectors = [
   '[contenteditable]',
 ]
-const inputTypeNeedSingleValueChangeRe = /^(date|time|month|week)$/
+const focusableWhenNotDisabledSelectors = [
```
rename

brian-mann commented on packages/driver/src/dom/selection.ts:
```diff
@@ -128,130 +118,105 @@ const getHostContenteditable = function (el) {
+//   }
+
+//   throw new Error('No selection in document')
+// }
```
commented code

brian-mann reviewed (commented):
The following fixes are not annotated anywhere in the code, or in the tests that test that they're working. Please update the source code + tests to link out to these issues.

fix #4587 fix not borrowing property getter for maxLength during type validation
fix #2166 allow targeting any focusable element during cy.type
fix #2613 not updating value in datetime-local input
fix #4767 allow typing minus(-) sign into number field if entire field is selected

brian-mann reviewed (changes requested):
See my comment here: https://github.com/cypress-io/cypress/pull/4870#pullrequestreview-305447476

brian-mann commented on packages/driver/test/cypress/integration/commands/actions/click_spec.js:
```diff
@@ -378,6 +378,49 @@ describe('src/cy/commands/actions/click', () => {
       cy.getAll('div', 'pointerover pointerenter pointerdown mousedown pointerup mouseup click').each(shouldBeCalled)
     })
 
+    it('events when element removed on pointerdown', () => {
```
@Bkucera why is this code here? this feels like code from the other `click` PR and not type related?

Bkucera replied:
snuck in during merge, fixed

kuceb committed:
link fixed issues in tests, actually fix one issue

kuceb committed:
cleanup remove dead code

Bkucera commented:
@brian-mann added links to issues in the tests, used #3661 instead of #2116

Bkucera commented on packages/driver/test/cypress/integration/commands/waiting_spec.coffee:
```diff
@@ -632,6 +632,7 @@ describe "src/cy/commands/waiting", ->
           .wait("@getUsers").then (xhr) ->
             expect(xhr.url).to.include "/users?num=4"
             expect(xhr.responseBody).to.deep.eq resp
+      null
```
this is to make eslint OK with skipping the last test in suite

kuceb committed:
remove unused helper

kuceb committed:
add test for #2166

kuceb committed:
pay no attention to the code and implementation changing... add tests and note for changing focus into date input

kuceb committed:
mark fix for issue #3316

kuceb committed:
mark fix for issue #3722

brian-mann commented on packages/driver/src/cypress/cy.coffee:
```diff
@@ -82,8 +82,8 @@ create = (specWindow, Cypress, Cookies, state, config, log) ->
   focused = $Focused.create(state)
-  keyboard = $Keyboard.create(state)
-  mouse = $Mouse.create(state, keyboard, focused)
+  keyboard = new $Keyboard(state)
```
go back to to using `create`

brian-mann commented on packages/driver/src/dom/selection.ts:
```diff
@@ -274,26 +229,22 @@ const moveCursorLeft = function (el) {
 
-  if ($elements.isContentEditable(el)) {
-    const selection = _getSelectionByEl(el)
+  // if ($elements.isContentEditable(el)) {
```
comment

brian-mann commented on packages/driver/src/dom/selection.ts:
```diff
@@ -274,26 +229,22 @@ const moveCursorLeft = function (el) {
+    'backward',
+    'character'
+  )
+  // }
```
comment

brian-mann commented on packages/driver/src/dom/selection.ts:
```diff
@@ -419,77 +379,111 @@ const selectAll = function (el) {
+    if (toStart) {
+      cursorPosition = 0
+    } else {
+      cursorPosition = $elements.getNativeProp(el, 'value').length
```
tighten up

brian-mann commented on packages/driver/test/cypress/integration/commands/actions/type_spec.js:
```diff
@@ -785,16 +771,13 @@ describe('src/cy/commands/actions/type', () => {
         .invoke('html', 'foobar')
-        .type('{selectAll}{rightarrow}{del}')
-        .then(() => {
+        .type('{movetoend}')
```
this is new functionality right? this needs a referenced issue

brian-mann commented on packages/driver/test/cypress/integration/commands/actions/type_spec.js:
```diff
@@ -167,6 +167,18 @@ describe('src/cy/commands/actions/type', () => {
+        cy.$$('input:first').focus()
+      })
+
+      cy.get('div[tabindex]:first').type('foobar')
```
what is the subject here? did it change to the new element? or stay as the old one?

brian-mann replied:
does not change

brian-mann commented on packages/driver/test/cypress/integration/commands/actions/type_spec.js:
```diff
@@ -1518,6 +1667,43 @@ describe('src/cy/commands/actions/type', () => {
+        it('follows focus into date input', () => {
+          let charCount = 0
+
+          cy.$$('input:first').on('input', (e) => {
```
`_.after(3, () => {})`

brian-mann commented on packages/driver/test/cypress/integration/commands/actions/type_spec.js:
```diff
@@ -1518,6 +1667,43 @@ describe('src/cy/commands/actions/type', () => {
+
+          let charCount = 0
+
+          cy.$$('input:first').on('input', (e) => {
```
`_.after(3, () => {})`

brian-mann commented on packages/driver/test/cypress/integration/commands/actions/type_spec.js:
```diff
@@ -2822,6 +3018,10 @@ describe('src/cy/commands/actions/type', () => {
       })
+
+      it('{shift} does not capitalize characters', () => {
+        cy.get('input:first').type('{shift}foo').should('have.value', 'foo')
```
likely need a note here to describe our implementation logic of why we consider this to be correct

brian-mann commented on packages/driver/test/cypress/integration/commands/actions/type_spec.js:
```diff
@@ -4373,13 +4601,15 @@ https://on.cypress.io/type`)
-          cy
-          .get(':text:first').type(val)
-          .should('have.value', val)
+      describe('naughtly strings', () => {
```
naughtly lol

brian-mann commented on packages/driver/src/cy/keyboard.ts:
```diff
@@ -47,680 +105,1010 @@ const toModifiersEventOptions = (modifiers) => {
-          return chars
-        })
-      }
+    throw Error(`Not a valid key: ${key}`)
```
i'm not sure this error can ever be hit

Bkucera replied:
It cannot

brian-mann commented on packages/driver/src/cy/keyboard.ts:
```diff
@@ -47,680 +105,1010 @@ const toModifiersEventOptions = (modifiers) => {
+          options.onValueChange(curText, el)
+        } else {
+          // key.events.textInput = false
         }
```
weird commented out `else` condition

brian-mann committed:
rename elements.isType -> elements.isInputType for clarity

brian-mann committed:
restructure keyboard a bit, don't export default, rework static class properties to regular exports

- attach $Mouse + $Keyboard to Cypress

brian-mann committed:
tighten up

brian-mann committed:
coffeescript fixes, consistent regexp const naming

brian-mann committed:
capitalize interface, tighten up several modifier functions

brian-mann committed:
misc cleanup and tightening up

brian-mann committed:
rework impl to use stubs instead of numbers

brian-mann commented on packages/driver/src/cy/keyboard.ts:
```diff
@@ -47,680 +107,1010 @@ const toModifiersEventOptions = (modifiers) => {
 
-        default: null
+    const modifierKeys = _.filter(keyDetailsArr, (key) => {
+      return isModifier(key)
```
tighten up

brian-mann committed:
some more cleanup

brian-mann committed:
fix logic with when function is called
